src/du_parameters/frequency.py

```diff
@@ -0,0 +1,526 @@
+# Copyright 2025 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""Define Frequency, GSCN and ARFCN classes by overloading arithmetic operators.
+
+Handle arithmetic operations with different types of objects.
+"""
+
+import logging
+from dataclasses import dataclass
+from decimal import Decimal
+
+logger = logging.getLogger(__name__)
+
+CONFIG_CONSTANT_THREE = Decimal("3")
+CONFIG_CONSTANT_TWO = Decimal("2")
+
+
+class GetRangeFromFrequencyError(Exception):
+    """Exception raised when frequency is not appropriate for any frequency range."""
+
+    pass
+
+
+class GetRangeFromGSCNError(Exception):
+    """Exception raised when GSCN is not appropriate for any frequency range."""
+
+    pass
+
+
+class Frequency(Decimal):
+    """Represents a frequency with unit conversion methods and arithmetic capabilities."""
+
+    class Units:
+        """Define Units for Frequency class."""
+
+        KHZ = 1000
+        MHZ = 1000 * 1000
+
+    @classmethod
+    def from_khz(cls, value: int | str) -> "Frequency":
+        """Convert KHz to Frequency."""
+        return cls(Decimal(value) * cls.Units.KHZ)
+
+    @classmethod
+    def from_mhz(cls, value: int | str) -> "Frequency":
+        """Convert MHz to Frequency."""
+        return cls(Decimal(value) * cls.Units.MHZ)
+
+    def __add__(self, other: "Frequency | Decimal | str | int") -> "Frequency":
+        """Add Frequency to others.
+
+        Args:
+            other (Frequency | Decimal | str | int): The value to add.
+
+        Returns:
+            Frequency: The resulting Frequency after the addition.
+        """
+        if not isinstance(other, (Frequency, Decimal, int, str)):
+            raise NotImplementedError(f"Unsupported type for addition: {type(other)}")
+        return Frequency(super().__add__(Decimal(other)))
+
+    def __sub__(self, other: "Frequency | Decimal | str | int") -> "Frequency":
+        """Subtract Frequency from others.
+
+        Args:
+            other (Frequency | Decimal | str | int): The value to subtract.
+
+        Returns:
+            Frequency: The resulting Frequency after the subtraction.
+        """
+        if not isinstance(other, (Frequency, Decimal, int, str)):
+            raise NotImplementedError(f"Unsupported type for subtraction: {type(other)}")
+        return Frequency(super().__sub__(Decimal(other)))
+
+    def __rsub__(self, other: "Frequency | Decimal | str | int") -> "Frequency":
+        """Subtract Frequency from others.
+
+        Args:
+            other (Frequency | Decimal | str | int): The value to reversely subtract from.
+
+        Returns:
+            Frequency: The resulting Frequency after the subtraction.
+        """
+        if not isinstance(other, (Frequency, Decimal, int, str)):
+            raise NotImplementedError(f"Unsupported type for subtraction: {type(other)}")
+        return Frequency(super().__sub__(Decimal(other)))
```

Comment:
This is a bug, `__rsub__` is called on the object on the right side of the operator. When you call `a - b`, Python will in most cases call `a.__sub__(b)` (unless b is a subclass of a, where it will call `b.__rsub__(a)` first instead).

When a does not know how to execute the operation on b's type, it raises `NotImplementedError`, and Python will give b a chance to execute the operation by calling `b.__rsub__(a)`.

In this case, if we run `a - b` and Python calls `b.__rsum__(a)`, the result will be for `b - a`, which is wrong.

```suggestion
        return Frequency(super().__rsub__(Decimal(other)))
```

A good way to test for this is this:

```
a = Frequency.from_khz(1)  # 1000
b = Decimal(500)           # 500

print(a - b)   # result should be: 500
print(b - a)   # result should be: -500
```